### PR TITLE
The Witness: Panel Hunt Mode

### PR DESCRIPTION
**Pull request by NewSoupVi**
Solve a specified number of random, visually marked panels in the game.

Features:
- Victory Condition: Panel Hunt
- Configurable amount of Total Panel Hunt Panels & Required Percentage
- Area hints have been adjusted to include that info

System Changes:
- Changed Event system to allow multiple events for the same entity

Tested:
I have played over a dozen seeds myself and we had a playtest.
I also looked at spoiler logs and all that.

### CODE REVIEW

File: worlds/witness/player_logic.py [hunk 1]
```diff
@@ -226,6 +228,10 @@ def make_single_adjustment(self, adj_type: str, line: str) -> None:
                     if entity_hex in self.DOOR_ITEMS_BY_ID and item_name in self.DOOR_ITEMS_BY_ID[entity_hex]:
                         self.DOOR_ITEMS_BY_ID[entity_hex].remove(item_name)
 
+        if adj_type == "Forbidden Doors":
```
Comment (Exempt-Medic) [inline worlds/witness/player_logic.py:231]: Are this and the other uses of it supposed to be here?

File: worlds/witness/player_logic.py [hunk 2]
```diff
@@ -387,6 +394,41 @@ def handle_postgame(self, world: "WitnessWorld") -> List[List[str]]:
...
+            self.HUNT_ENTITIES.add("0x09F7F")
+            self.COMPLETELY_DISABLED_ENTITIES.add("0x3D9A9")  # Elevator Start
+
+            # If mountain lasers are disabled, and challenge lasers > 7, the box will need to be rotated
+            if chal_lasers > 7:
+                postgame_adjustments.append([
+                    "Requirement Changes:",
+                    "0xFFF00 - 11 Lasers - True",
+                ])
+        if disable_challenge_lasers:
```
Comment (Exempt-Medic) [inline worlds/witness/player_logic.py:425]: You usually put an extra line in places like this
```suggestion

        if disable_challenge_lasers:
```

File: worlds/witness/player_logic.py [hunk 3]
```diff
@@ -401,7 +443,11 @@ def make_options_adjustments(self, world: "WitnessWorld") -> None:
 
         # Exclude panels from the post-game if shuffle_postgame is false.
         if not world.options.shuffle_postgame:
```
Comment (Exempt-Medic) [inline worlds/witness/player_logic.py:445]: Does this matter if the goal is panel_hunt / should this be `if not (world.options.shuffle_postgame or victory == "panel_hunt")`

File: worlds/witness/player_logic.py [hunk 4]
```diff
@@ -794,23 +846,124 @@ def determine_unrequired_entities(self, world: "WitnessWorld") -> None:
...
-    def make_event_item_pair(self, entity_hex: str) -> Tuple[str, str]:
-        """
-        Makes a pair of an event panel and its event item
-        """
-        action = " Opened" if self.REFERENCE_LOGIC.ENTITIES_BY_HEX[entity_hex]["entityType"] == "Door" else " Solved"
+    def pick_panel_hunt_panels(self, world: "WitnessWorld"):
+        disallowed_entities_for_panel_hunt = {
+            "0x03629",  # Tutorial Gate Open, which is the panel that is locked by panel hunt
+            "0x03505",  # Tutorial Gate Close (same thing)
+            "0x3352F",  # Gate EP (same thing)
```
Comment (Exempt-Medic) [inline worlds/witness/player_logic.py:864]: Does this need to be here, it's not a panel?

File: worlds/witness/player_logic.py [hunk 5]
```diff
@@ -794,23 +846,124 @@ def determine_unrequired_entities(self, world: "WitnessWorld") -> None:
...
+                    area: (max_percentage - current_percentage) / max_percentage
+                    for area, current_percentage in contributing_percentage_per_area.items()
+                }
+                # use same_area_discouragement as lerp factor
+                allowance_per_area = {
+                    area: (1.0 - same_area_discouragement_factor) + (weight * same_area_discouragement_factor)
+                    for area, weight in allowance_per_area.items()
+                }
+
+            assert min(allowance_per_area.values()) >= 0, (f"Somehow, an area had a negative weight when picking"
```
Comment (Exempt-Medic) [inline worlds/witness/player_logic.py:933]: Do you want this to be an `assert` or something like an exception so it'll still error when built?

**Review by Exempt-Medic — CHANGES_REQUESTED** [review]
The code changes LGTM, some questions were clarified, a few minor things and questions below. Did some testing with generations but will do some more in-depth testing with particular edge-cases to test

File: worlds/witness/hints.py [hunk 6]
```diff
@@ -583,12 +602,28 @@ def create_all_hints(world: "WitnessWorld", hint_amount: int, area_hints: int,
...
+    if location and location.address is not None:
+        arg_1 = location.address
+    elif hint.area is not None:
+        arg_1 = hint.area
+    else:
+        arg_1 = -1
+
+    # self.player if junk hint, player if location hint, progression amount & hunt panels if area hint
+    arg_2: int
+    if area_amount is not None:
```
[the remark this answers is not in the document]
Comment (NewSoupVi) [reply worlds/witness/hints.py:618]: Then area_amount is **0**, not None

File: worlds/witness/player_logic.py (hunk 1 above)
Comment (NewSoupVi) [reply worlds/witness/player_logic.py:231]: Uh oh stinky merge error

File: worlds/witness/player_logic.py (hunk 4 above)
Comment (NewSoupVi) [reply worlds/witness/player_logic.py:861]: Generic Entity Hunt foresight

File: worlds/witness/player_logic.py (hunk 3 above)
Comment (NewSoupVi) [reply worlds/witness/player_logic.py]: True

File: worlds/witness/player_logic.py (hunk 5 above)
Comment (NewSoupVi) [reply worlds/witness/player_logic.py:930]: I prefer the assert here, this is more for me to notice that it happened than to crash generation - A negative weight is treated as 0, so I'd be crashing a generation that would be successful otherwise

**Comment by agilbert1412** [comment]
Since #2698 is merged, you should probably update the description to not say "don't review" lmao

File: worlds/witness/data/static_logic.py [hunk 7]
```diff
@@ -103,19 +104,31 @@ def read_logic_file(self, lines) -> None:
                 "Laser Hedges",
                 "Laser Pressure Plates",
             }
-            is_vault_or_video = "Vault" in entity_name or "Video" in entity_name
+            is_vault_or_video = "Vault" in entity_name
```
Comment (agilbert1412) [inline worlds/witness/data/static_logic.py:107]: if this no longer checks for video, the variably should probably be renamed

File: worlds/witness/hints.py [hunk 8]
```diff
@@ -386,22 +387,22 @@ def get_hintable_areas(world: "WitnessWorld") -> Tuple[Dict[str, List[Location]]
...
+def word_area_hint(world: "WitnessWorld", hinted_area: str, area_items: List[Item]) -> Tuple[str, int, Optional[int]]:
     """
     Word the hint for an area using natural sounding language.
     This takes into account how much progression there is, how much of it is local/non-local, and whether there are
     any local lasers to be found in this area.
     """
 
-    local_progression = sum(item.player == world.player and item.advancement for item in corresponding_items)
-    non_local_progression = sum(item.player != world.player and item.advancement for item in corresponding_items)
+    local_progression = sum(item.player == world.player and item.advancement for item in area_items)
```
Comment (agilbert1412) [inline worlds/witness/hints.py:397]: I suspect you could get slightly better performance by turning this into a loop, to iterate only once and check the advancement condition less
```py
local_progression = 0
non_local_progression = 0
for item in area_items:
    if not item.advancement:
        continue
    if item.player == world.player:
        local_progression += 1
    else:
        non_local_progression += 1
```

But also, I don't personally micro optimizations like that are relevant, and it makes me incomfortable that I am apparently at the stage that I even thought of it. So, if you want, you can do us a both a favor and ignore this comment :)

File: worlds/witness/hints.py [hunk 9]
```diff
@@ -583,12 +602,28 @@ def create_all_hints(world: "WitnessWorld", hint_amount: int, area_hints: int,
...
-
-    # None if junk hint, address if location hint, area string if area hint
-    arg_1 = location.address if location else (hint.area if hint.area else None)
-
-    # self.player if junk hint, player if location hint, progression amount if area hint
-    arg_2 = area_amount if area_amount is not None else (location.player if location else local_player_number)
+    hunt_panels = hint.area_hunt_panels
+
+    # -1 if junk hint, address if location hint, area string if area hint
+    arg_1: Union[str, int]
```
Comment (agilbert1412) [inline worlds/witness/hints.py:608]: I'm not a big fan of these vague and generic variable names. I understand they can hold multiple different things, but here's what I would do

Turn the whole first arg code into a different method, where the arg_1 is actually multiple variables, or even better, no variable, with just some "return [whatever]" at the relevant lines. Then do the same for the arg_2 code.
Then, here, you can also not use the variables at all, and just return the method calls

But that's definitely not critical, feel free to ignore

File: worlds/witness/player_logic.py [hunk 10]
```diff
@@ -387,6 +390,42 @@ def handle_postgame(self, world: "WitnessWorld") -> List[List[str]]:
...
+            or panel_hunt_postgame == "disable_anything_locked_by_lasers"
+        )
+
+        disable_challenge_lasers = (
+            panel_hunt_postgame == "disable_challenge_lasers_locations"
+            or panel_hunt_postgame == "disable_anything_locked_by_lasers"
+        )
+
+        if disable_mountain_lasers:
+            self.DISABLE_EVERYTHING_BEHIND.add("0x09F7F")  # Short box
```
Comment (agilbert1412) [inline worlds/witness/player_logic.py:411]: If you're writing comments to explain the meaning of the hexadecimal magic number, why not turn it into a named constant instead?

Comment is applicable to many places throughout this file, and probably a few more files

File: worlds/witness/player_logic.py [hunk 11]
```diff
@@ -794,23 +843,150 @@ def determine_unrequired_entities(self, world: "WitnessWorld") -> None:
...
         self.ENTITIES_WITHOUT_ENSURED_SOLVABILITY |= {
             item_name for item_name, is_required in is_item_required_dict.items() if not is_required
         }
 
-    def make_event_item_pair(self, entity_hex: str) -> Tuple[str, str]:
-        """
-        Makes a pair of an event panel and its event item
-        """
-        action = " Opened" if self.REFERENCE_LOGIC.ENTITIES_BY_HEX[entity_hex]["entityType"] == "Door" else " Solved"
+    def pick_panel_hunt_panels(self, world: "WitnessWorld"):
```
Comment (agilbert1412) [inline worlds/witness/player_logic.py:857]: A lot of sections in this algorithm are explained with a comment. That's usually a sign that it could be split into more methods, with explicit names, and it would be more readable and more maintainable

File: worlds/witness/rules.py [hunk 12]
```diff
@@ -225,24 +228,26 @@ def _has_item(item: str, world: "WitnessWorld", player: int,
     if item in player_logic.REFERENCE_LOGIC.ALL_REGIONS_BY_NAME:
         region = world.get_region(item)
         return region.can_reach
-    if item == "7 Lasers":
+    elif item == "7 Lasers":
```
Comment (agilbert1412) [inline worlds/witness/rules.py:231]: Why turn all of these into elif, if every block has a `return` anyway?

**Review by agilbert1412 — APPROVED** [review]
Reviewed the entirety of the code and left a handful of suggestions, but I don't believe any of these suggestions is critical, and they are all subjective (As in, Violet could decide to disagree and that would be reasonable). So I approve with suggestions

File: worlds/witness/player_logic.py (hunk 10 above)
Comment (NewSoupVi) [reply worlds/witness/player_logic.py:411]: Hmmm, I'd have to auto generate the constants. I guess that's possible? Like can you make an enum class from a dict? Probably I assume

It's definitely omega out of scope for this PR tho since this is in like *every file* lol

File: worlds/witness/rules.py (hunk 12 above)
Comment (NewSoupVi) [reply worlds/witness/rules.py:231]: True

File: worlds/witness/hints.py (hunk 8 above)
Comment (NewSoupVi) [reply worlds/witness/hints.py:397]: LMAO Yeah that makes sense. I do like optimised code but I do also like short code... So as long as your code is either of those, and you don't write unoptimised long code :D

File: worlds/witness/data/static_logic.py (hunk 7 above)
Comment (NewSoupVi) [reply worlds/witness/data/static_logic.py:107]: True

File: worlds/witness/player_logic.py (hunk 10 above)
Comment (agilbert1412) [reply worlds/witness/player_logic.py:411]: I was suggesting the really low-effort version, like
```py
SHORT_BOX = "0x09F7F"
# [...]
if disable_mountain_lasers:
    self.DISABLE_EVERYTHING_BEHIND.add(SHORT_BOX)
```
But if you have something cleaner and smarter in mind, go right ahead lmao

**Comment by NewSoupVi** [comment]
Kaito's comment about the function being way too big made a lot of sense to me. The best way I found to address this is to make an entire new class called `EntityHuntPicker`

It's a lot of deleted and added lines, but I've only moved stuff around. It's obviously possible that I missed stuff in the process.

**Review by agilbert1412 — APPROVED** [review]
I think this is a greatly improved now

**Review by agilbert1412 — APPROVED** [review]

**Review by Exempt-Medic — APPROVED** [review]
The changes LGTM. Looked over all the code but more carefully stepped through the postgame considerations and `player_logic.py` and read through `options.py`. I think some of the requirements changes are a little weird, but they all work just fine. Tested various option configurations and checked through spoiler logs to ensure the disabled entities made sense. Printed out a lot of hint strings and read them over for proper formatting.

File: worlds/witness/options.py
Comment (Exempt-Medic) [inline worlds/witness/options.py:1]: There's inconsistent usage of "Hunt Panels" "hunt panels" "Panel Hunt panels" and other ways of saying this same concept. Unsure if you want to "fix" that though

Comment (NewSoupVi) [reply worlds/witness/options.py:1]: I am okay with "Panel Hunt panels" vs "Hunt Panels" (verbose vs shortened), but I did change all "hunt panels" to "Hunt Panels" now.

File: worlds/witness/entity_hunt.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        """
+
+        replacements = {
+            "0x18488": "0x00609",  # Replace Swamp Sliding Bridge Underwater -> Swamp Sliding Bridge Above Water
+            "0x03676": "0x03678",  # Replace Quarry Upper Ramp Control -> Lower Ramp Control
+            "0x03675": "0x03679",  # Replace Quarry Upper Lift Control -> Lower Lift Control
+
+            "0x03702": "0x15ADD",  # Jungle Vault Box -> Jungle Vault Panel
+            "0x03542": "0x002A6",  # Mountainside Vault Box -> Mountainside Vault Panel
+            "0x03481": "0x033D4",  # Jungle Vault Box -> Jungle Vault Panel
```
Comment (Exempt-Medic) [inline worlds/witness/entity_hunt.py:193]: Also, shouldn't this include Shipwreck?
```suggestion
            "0x03481": "0x033D4",  # Tutorial Vault Box -> Tutorial Vault Panel
```

**Comment by NewSoupVi** [comment]
Drafting this because I want to add unit tests for this.

**Comment by NewSoupVi** [comment]
Added some unit tests :)

**Review by Berserker66 — APPROVED** [review]
